modules/hcp-ec2-client/templates/hashicups.nomad

```diff
@@ -25,35 +29,90 @@ variable "product_db_port" {
 
 job "hashicups" {
   datacenters = ["dc1"]
-
-  group "frontend" {
+  group "nginx" {
     network {
       mode = "bridge"
 
-      port "http" {
-        static = var.frontend_port
+      port "nginx" {
+        static = var.nginx_port
       }
     }
 
     service {
-      name = "frontend"
-      port = "http"
-    }
-
-    task "frontend" {
-      driver = "docker"
-
-      config {
-        image = "hashicorpdemoapp/frontend:v1.0.2"
-        ports = ["http"]
+      name = "nginx"
+      port = "nginx"
+      connect {
+        sidecar_service {
+          proxy {
+            upstreams {
+              destination_name = "frontend"
+              local_bind_port  = var.frontend_port
+            }
+            upstreams {
+              destination_name = "public-api"
+              local_bind_port  = var.public_api_port
+            }
+          }
+        }
       }
+    }
 
-      env {
-        NEXT_PUBLIC_PUBLIC_API_URL = "/"
+    #remove caching from nginx. It makes it easier to show things like blue green deployments/canary. Otherwise nginx caches too much. 
+    task "nginx" {
+        driver = "docker"
+        meta {
+          service = "nginx-reverse-proxy"
+        }
+        config {
+          image = "nginx:alpine"
+          ports = ["nginx"]
+          mount {
+            type   = "bind"
+            source = "local/default.conf"
+            target = "/etc/nginx/conf.d/default.conf"
+          }
+        }
+        resources {
+        cpu = 300 # MHz
+        memory = 128 # MB
+       }
+        template {
+          data =  <<EOF
+            proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=STATIC:10m inactive=7d use_temp_path=off;
+            upstream frontend_upstream {
+              server {{ env "NOMAD_UPSTREAM_ADDR_frontend" }};
+            }
+            upstream public_api_upstream {
+              server {{ env "NOMAD_UPSTREAM_ADDR_public_api" }};
+            }
+            server {
+              listen {{ env "NOMAD_PORT_nginx" }};
+              server_name {{ env "NOMAD_IP_nginx" }};
+              server_tokens off;
+              gzip on;
+              gzip_proxied any;
+              gzip_comp_level 4;
+              gzip_types text/css application/javascript image/svg+xml;
+              proxy_http_version 1.1;
+              proxy_set_header Upgrade $http_upgrade;
+              proxy_set_header Connection 'upgrade';
+              proxy_set_header Host $host;
+             
+              
+              location / {
+                proxy_pass http://frontend_upstream/;
+                add_header x-debug 2;
+              } 
+                location /api {
+                proxy_pass http://public_api_upstream;
+              }
+            
+          }
+          EOF
+          destination = "local/default.conf"
       }
     }
   }
-
```

Comment:
nit: can we leave all these newlines between the nomad stanzas? Imo they help a bit with readability